vpr/src/device/rr_graph/rr_graph_obj.cpp

```diff
@@ -0,0 +1,1753 @@
+        return RRGRAPH_OPEN_NODE_ID;
+    }
+
+    /* Check if x, y, type and ptc, side is valid */
+    if ((ptc < 0)                                                 /* See if ptc is smaller than the index of first element */
+        || (size_t(ptc) > node_lookup_[x][y][type].size() - 1)) { /* See if ptc is large than the index of last element */
+        /* Return a zero range! */
+        return RRGRAPH_OPEN_NODE_ID;
+    }
+
+    /* Check if x, y, type and ptc, side is valid */
+    /* iside is always larger than -1, we can skip checking */
+    if (iside > node_lookup_[x][y][type][ptc].size() - 1) { /* See if side is large than the index of last element */
+        /* Return a zero range! */
+        return RRGRAPH_OPEN_NODE_ID;
+    }
+
+    return node_lookup_[x][y][itype][ptc][iside];
+}
+
+RRGraph::node_range RRGraph::find_nodes(short x, short y, t_rr_type type, int ptc) const {
+    if (!valid_fast_node_lookup()) {
+        build_fast_node_lookup();
+    }
+
+    /* IMPORTANT!!! 
+     * check x, y, type, ptc if they are in the range
+     * if not, return a zero range 
+     */
+    size_t itype = type;
+
+    /* Check if x, y, type and ptc, side is valid */
+    if ((x < 0)                                     /* See if x is smaller than the index of first element */
+        || (size_t(x) > node_lookup_.size() - 1)) { /* See if x is large than the index of last element */
+        /* Return a zero range! */
+        const std::vector<RRNodeId> zero_range = std::vector<RRNodeId>();
+        return vtr::make_range(zero_range.begin(), zero_range.end());
+    }
+
+    /* Check if x, y, type and ptc, side is valid */
+    if ((y < 0)                                        /* See if y is smaller than the index of first element */
+        || (size_t(y) > node_lookup_[x].size() - 1)) { /* See if y is large than the index of last element */
+        /* Return a zero range! */
+        const std::vector<RRNodeId> zero_range = std::vector<RRNodeId>();
+        return vtr::make_range(zero_range.begin(), zero_range.end());
+    }
+
+    /* Check if x, y, type and ptc, side is valid */
+    /* itype is always larger than -1, we can skip checking */
+    if (itype > node_lookup_[x][y].size() - 1) { /* See if type is large than the index of last element */
+        /* Return a zero range! */
+        const std::vector<RRNodeId> zero_range = std::vector<RRNodeId>();
+        return vtr::make_range(zero_range.begin(), zero_range.end());
+    }
+
+    /* Check if x, y, type and ptc, side is valid */
+    if ((ptc < 0)                                                 /* See if ptc is smaller than the index of first element */
+        || (size_t(ptc) > node_lookup_[x][y][type].size() - 1)) { /* See if ptc is large than the index of last element */
+        /* Return a zero range! */
+        const std::vector<RRNodeId> zero_range = std::vector<RRNodeId>();
+        return vtr::make_range(zero_range.begin(), zero_range.end());
```

Comment:
Another thing to consider is whether specifying an out-of-range value should be accepted (and a zero-sized range returned), or if it should be a hard error (i.e. throw an exception with `VPR_ERROR()`).

My preference would be to throw an error, otherwise a bug in the code calling this function would go unnoticed.

Reply:
I have removed the function. And we will see if we do need the function later.